src/module/migration/migrations/921-kineticist-restructure.ts

```diff
@@ -0,0 +1,201 @@
+                await this.#setElement(source, elementFour, "four");
+            }
+        }
+
+        const thirdThreshold = source.items.find(
+            (i) => (i.system.slug ?? sluggify(i.name)) === "third-gates-threshold",
+        );
+
+        if (thirdThreshold) {
+            const elementFive = thirdThreshold.system.rules.find(
+                (r): r is ChoiceSetSource => r.key === "ChoiceSet" && "flag" in r && r.flag === "elementFork",
+            )?.selection;
+
+            if (typeof elementFive === "string") {
+                await this.#setElement(source, elementFive, "five");
+            }
+        }
+
+        const fourthThreshold = source.items.find(
+            (i) => (i.system.slug ?? sluggify(i.name)) === "fourth-gates-threshold",
+        );
+
+        if (fourthThreshold) {
+            const elementSix = fourthThreshold.system.rules.find(
+                (r): r is ChoiceSetSource => r.key === "ChoiceSet" && "flag" in r && r.flag === "elementFork",
+            )?.selection;
+
+            if (typeof elementSix === "string") {
+                await this.#setElement(source, elementSix, "six");
+            }
+        }
+    }
+
+    async #loadFeatSource(uuid: ItemUUID): Promise<FeatSource | null> {
+        const item = await fromUuid<ItemPF2e>(uuid);
+        const source: ItemSourcePF2e | null = item?.toObject(true) ?? null;
+        return source && itemIsOfType(source, "feat") ? source : null;
+    }
+
+    #setGate(gate: FeatSource, gateNumber: string, element: string) {
+        const modifiedRules = [
+            {
+                key: "ActiveEffectLike",
+                mode: "override",
+                path: "flags.pf2e.kineticist.gate." + gateNumber,
+                value: element,
+            },
+        ];
+
+        gate.system.rules.push(...modifiedRules);
+
+        switch (gateNumber) {
+            case "three":
+                gate.system.level.value = 5;
+                break;
+            case "four":
+                gate.system.level.value = 9;
+                break;
+            case "five":
+                gate.system.level.value = 13;
+                break;
+            case "six":
+                gate.system.level.value = 17;
+                break;
+        }
+
+        gate.system.rules.splice(0, 2);
+    }
+
+    async #setElement(source: ActorSourcePF2e, element: string, gateNumber: string) {
+        const elementMap: Map<string, ItemUUID> = new Map([
+            ["air", "Compendium.pf2e.classfeatures.Item.X11Y3T1IzmtNqGMV"],
+            ["earth", "Compendium.pf2e.classfeatures.Item.dEm00L1XFXFCH2wS"],
+            ["fire", "Compendium.pf2e.classfeatures.Item.PfeDtJBJdUun0THS"],
+            ["metal", "Compendium.pf2e.classfeatures.Item.21JjdNW0RQ2LfaH3"],
+            ["water", "Compendium.pf2e.classfeatures.Item.MvunDFH8Karxee0t"],
+            ["wood", "Compendium.pf2e.classfeatures.Item.8X8db58vKx21L0Dr"],
+        ]);
```

Comment:
pull this out as a separate property instead of reconstructing it every time

Reply:
Good now?